File: src/main/services/osSandboxRunner.ts
import os from 'os'
import { spawn } from 'child_process'
import { normAbs } from './acpCore'

export type OsSandboxPermissionMode = 'workspace_whitelist' | 'full_access'

export type OsSandboxRunOptions = {
  command: string
  args?: string[]
  cwd: string
  workspaceDir: string
  permissionMode: OsSandboxPermissionMode
  timeoutMs?: number
  allowedRoots?: string[]
  env?: NodeJS.ProcessEnv
}

export type OsSandboxRunResult = {
  code: number | null
  signal: NodeJS.Signals | null
  stdout: string
  stderr: string
  sandbox: {
    enabled: boolean
    kind: 'none' | 'macos_sandbox_exec'
    reason: string
  }
}

function escapeProfileLiteral(v: string): string {
  return String(v || '').replaceAll('\\', '\\\\').replaceAll('"', '\\"')
}

function uniqueRoots(values: string[]): string[] {
  const out: string[] = []
  const seen = new Set<string>()
  for (const raw of values) {
    const s = String(raw || '').trim()
    if (!s) continue
    const p = normAbs(s)
    if (seen.has(p)) continue
    seen.add(p)
    out.push(p)
  }
  return out
}

function buildMacSandboxProfile(writeRoots: string[]): string {
  const roots = uniqueRoots(writeRoots)
  const rules = roots
    .map((root) => {
      const s = escapeProfileLiteral(root)
      return `(allow file-read* (literal "${s}") (subpath "${s}"))
(allow file-write* (literal "${s}") (subpath "${s}"))`
    })
    .join('\n')
  return `(version 1)
(deny default)
(import "system.sb")
(allow process*)
(allow signal (target self))
(allow sysctl-read)
(allow file-read*
  (subpath "/System")
  (subpath "/usr")
  (subpath "/bin")
  (subpath "/sbin")
  (subpath "/private/etc")
  (subpath "/private/var")
  (subpath "/dev")
)
(deny network*)
${rules}
`
}

function clampTimeout(value: number | undefined): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) return 120_000
  return Math.max(0, value)
}

export async function runWithOsSandbox(opts: OsSandboxRunOptions): Promise<OsSandboxRunResult> {
  const command = String(opts.command || '').trim()
  if (!command) throw new Error('command is required')
  const args = Array.isArray(opts.args) ? opts.args.map((v) => String(v)) : []
  const cwd = normAbs(String(opts.cwd || '').trim())
  const workspaceRaw = String(opts.workspaceDir || '').trim()
  if (!workspaceRaw) throw new Error('workspaceDir is required')
  const workspaceDir = normAbs(workspaceRaw)
  const timeoutMs = clampTimeout(opts.timeoutMs)
  const env = { ...process.env, ...(opts.env || {}) }

  const writeRoots = uniqueRoots([workspaceDir, os.tmpdir(), ...(opts.allowedRoots || [])])
  const useSandbox = process.platform === 'darwin' && opts.permissionMode !== 'full_access'
  const spawnCommand = useSandbox ? 'sandbox-exec' : command
  const spawnArgs = useSandbox ? ['-p', buildMacSandboxProfile(writeRoots), command, ...args] : args
  const sandboxMeta = useSandbox
    ? { enabled: true, kind: 'macos_sandbox_exec' as const, reason: 'permission_mode_workspace_whitelist' }
    : {
        enabled: false,
        kind: 'none' as const,
        reason: opts.permissionMode === 'full_access' ? 'permission_mode_full_access' : `platform_${process.platform}`
      }

  const child = spawn(spawnCommand, spawnArgs, { cwd, env, shell: false })
  let stdout = ''
  let stderr = ''
  child.stdout.setEncoding('utf8')
  child.stderr.setEncoding('utf8')
  child.stdout.on('data', (d: string) => (stdout += d))
  child.stderr.on('data', (d: string) => (stderr += d))

  const res = await new Promise<{ code: number | null; signal: NodeJS.Signals | null }>((resolve, reject) => {
    let settled = false
    const settle = (cb: () => void) => {
      if (settled) return
      settled = true
      clearTimeout(t)
      cb()
    }
    const t = setTimeout(() => {
      try {
        child.kill('SIGTERM')
      } catch {
        settle(() => reject(new Error('Command timed out')))
        return
      }
      setTimeout(() => {
        try {
          child.kill('SIGKILL')
        } catch {
          return
        }
      }, 1500)
      settle(() => reject(new Error('Command timed out')))
    }, timeoutMs)

    child.on('exit', (code, signal) => settle(() => resolve({ code, signal })))
    child.on('error', (e: NodeJS.ErrnoException) => {
      if (useSandbox && e?.code === 'ENOENT') {
        settle(() => reject(new Error('sandbox-exec is not available on this system')))
        return
      }
      settle(() => reject(e))
    })
  })

  return { code: res.code, signal: res.signal, stdout, stderr, sandbox: sandboxMeta }
}

File: pybackend/anima_backend_shared/os_sandbox_runner.py
import os
import subprocess
import sys
import tempfile
from typing import Any, Dict, List, Optional

from .util import norm_abs, safe_env


def _escape_profile_literal(value: str) -> str:
    return str(value or "").replace("\\", "\\\\").replace('"', '\\"')


def _unique_roots(values: List[str]) -> List[str]:
    out: List[str] = []
    seen = set()
    for raw in values:
        s = str(raw or "").strip()
        if not s:
            continue
        p = norm_abs(s)
        if p in seen:
            continue
        seen.add(p)
        out.append(p)
    return out


def _build_macos_profile(write_roots: List[str]) -> str:
    rules: List[str] = []
    for root in _unique_roots(write_roots):
        s = _escape_profile_literal(root)
        rules.append(f'(allow file-read* (literal "{s}") (subpath "{s}"))')
        rules.append(f'(allow file-write* (literal "{s}") (subpath "{s}"))')
    return "\n".join(
        [
            "(version 1)",
            "(deny default)",
            '(import "system.sb")',
            "(allow process*)",
            "(allow signal (target self))",
            "(allow sysctl-read)",
            "(allow file-read*",
            '  (subpath "/System")',
            '  (subpath "/usr")',
            '  (subpath "/bin")',
            '  (subpath "/sbin")',
            '  (subpath "/private/etc")',
            '  (subpath "/private/var")',
            '  (subpath "/dev")',
            ")",
            "(deny network*)",
            *rules,
            "",
        ]
    )


def run_bash_with_os_sandbox(
    *,
    command: str,
    cwd: str,
    timeout_ms: int,
    permission_mode: str,
    workspace_dir: str,
    allowed_roots: Optional[List[str]] = None,
    env: Optional[Dict[str, str]] = None,
    max_chars: int = 20000,
) -> Dict[str, Any]:
    cmd = str(command or "").strip()
    if not cmd:
        raise RuntimeError("command is required")

    run_cwd = norm_abs(str(cwd or "").strip())
    timeout = max(1000, min(int(timeout_ms or 20000), 120000))
    base_env = dict(env) if isinstance(env, dict) else safe_env()

    write_roots = _unique_roots([workspace_dir, tempfile.gettempdir(), *(allowed_roots or [])])
    use_sandbox = sys.platform == "darwin" and str(permission_mode or "").strip() != "full_access"

    bash_base = ["/bin/bash", "--noprofile", "--norc", "-c", cmd]
    if use_sandbox:
        command_list = ["sandbox-exec", "-p", _build_macos_profile(write_roots), *bash_base]
        sandbox = {"enabled": True, "kind": "macos_sandbox_exec", "reason": "permission_mode_workspace_whitelist"}
    else:
        command_list = list(bash_base)
        sandbox = {
            "enabled": False,
            "kind": "none",
            "reason": "permission_mode_full_access" if str(permission_mode or "").strip() == "full_access" else f"platform_{sys.platform}",
        }

    p = subprocess.run(
        command_list,
        cwd=run_cwd,
        capture_output=True,
        text=True,
        env=base_env,
        timeout=timeout / 1000.0,
    )
    stdout = p.stdout or ""
    stderr = p.stderr or ""
    out_trunc = len(stdout) > max_chars
    err_trunc = len(stderr) > max_chars
    if out_trunc:
        stdout = stdout[:max_chars]
    if err_trunc:
        stderr = stderr[:max_chars]
    return {
        "ok": True,
        "exitCode": int(p.returncode),
        "stdout": stdout,
        "stderr": stderr,
        "truncated": {"stdout": out_trunc, "stderr": err_trunc},
        "cwd": run_cwd,
        "sandbox": sandbox,
    }
